polyply/src/random_walk.py

```diff
@@ -136,20 +325,57 @@ def _random_walk(meta_molecule):
         ----------
         meta_molecule:  :class:`polyply.src.meta_molecule.MetaMolecule`
         """
-        first_node = list(meta_molecule.nodes)[0]
+        if not self.start_node:
+            first_node = _find_starting_node(meta_molecule)
+        else:
+            first_node = self.start_node
+        if "position" not in meta_molecule.nodes[first_node]:
+            constrained = full_fill_geometrical_constraints(self.start, self.molecule.nodes[first_node])
+            if not self._is_overlap(self.start, first_node) and constrained:
+                self.nonbond_matrix.add_positions(self.start, self.mol_idx, first_node, start=True)
+                self.success = True
+            else:
+                self.success = False
+                return
 
-        if not "position" in meta_molecule.nodes[first_node]:
-            meta_molecule.nodes[first_node]["position"] = np.array([0, 0, 0])
+        vector_bundle = self.vector_sphere.copy()
+        count = 0
+        placed_nodes = []
+        path = list(nx.bfs_edges(meta_molecule, source=first_node))
+        step_count = 0
+        while step_count < len(path):
+            prev_node, current_node = path[step_count]
+            if not meta_molecule.nodes[current_node]["build"]:
+                step_count += 1
+                continue
+
+            status = self.update_positions(vector_bundle,
+                                           current_node,
+                                           prev_node)
+            self.success = status
+            placed_nodes.append((step_count, current_node))
 
-        vector_bundle = norm_sphere(5000)
+            # in principle this count here is not needed, however,
+            # we need to check the performance in terms of strucutre
+            # generation before doing any adjustments here
+            if not self.success and count < self.maxiter:
+                nrewind = 5
+                if len(placed_nodes) < nrewind+1:
+                   return
+                step_count = self._rewind(step_count, placed_nodes, nrewind)
+                placed_nodes = placed_nodes[:-nrewind]
```

Comment:
I would make rewind do this. Related, perhaps placed_nodes should be an instance variable